File: sdk_samples/samples/CSharp/04_multi_stream/04_multi_stream.cs
/**
 * CARMEN Video SDK
 *
 * @category    CSharp Sample
 * @package     CARMEN Video CSharp Sample
 *
 * @copyright   2024 Adaptive Recognition
 * @licence     https://adaptiverecognition.com/eula/
 *
 * This sample shows how you can set up a process with CARMEN Video SDK that
 * recognizes license plates (ANPR) and vehicle attributes (MMR) on multiple video streams.
 * There are "TODO" comments in the file where the code must be modified.
 */

using Carmen;
using Carmen.Anpr;
using Carmen.Mmr;
using Carmen.Video;

class Sample04_multi_stream
{
    public struct StreamConfig
    {
        public StreamConfig(string url, string region, string name)
        {
            Url = url;
            Region = region;
            Name = name;
        }

        public string Url { get; }
        public string Region { get; }
        public string Name { get; }
    }

//TODO: change them for valid streams and engines
    public static readonly List<StreamConfig> streamConfigs = new List<StreamConfig>
    {
        new StreamConfig("http://192.168.6.50:9901/video.mjpeg", "EUR", "Stream 1"),
        new StreamConfig("file:C:/Program Files/videos/video.mp4", "NAM", "Stream 2")
    };

//TODO: EventCallback is called from a StreamProcessor instance's dedicated "EventCallback executor" thread.
// If you are using the same resource in multiple StreamProcessors' EventCallback
// you may have to protect against race conditions

    static readonly object __lockEventCallback = new object ();

    static void EventHandlerCallback(Event e)
    {
        lock(__lockEventCallback){
            try
            {
                using (e)
                {
                    Console.WriteLine("------------------------------------------------------");
                    Console.WriteLine("Channel name: " + e.ChannelName);
                    Console.WriteLine("Unix timestamp: " + e.Timestamp + " ms (" + e.DateTime.ToLocalTime() + ")");

                    Console.WriteLine("Plate text: " + e.Vehicle.Plate.Text);
                    Console.WriteLine("Country: " + e.Vehicle.Plate.Country);

                    Console.WriteLine("Make: " + e.Vehicle.Attributes.Make);
                    Console.WriteLine("Model: " + e.Vehicle.Attributes.Model);
                    Console.WriteLine("Color: " + e.Vehicle.Attributes.Color);
                    Console.WriteLine("Category: " + e.Vehicle.Attributes.Category);

                    Console.WriteLine();
                }
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine("Exception in event callback. " + ex);
            }
        }
    }

    static void Main(string[] args)
    {
        try
        {
            // INITIALIZE COMMON ANPR OBJECT
            using Anpr.AnprBuilder anprBuilder = Anpr.Builder();
            using Anpr anpr = anprBuilder
                .Type(AnprType.Local) // use LocalGo for CARMEN_GO engines
                .LocalConcurrencyLimit(2) //TODO: change it to a value that matches the number of core licences you have
                .Build();

            // INITIALIZE COMMON MMR (Make and Model Recognition) OBJECT
            using Mmr.MmrBuilder mmrBuilder = Mmr.Builder();
            using Mmr mmr = mmrBuilder
                .Type(MmrType.Local)
                .Build();

            // BUILD STREAM PROCESSOR OBJECTS THAT USE THE COMMON ANPR AND MMR RESOURCES
            List<StreamProcessor> streams = streamConfigs.Select(stream => {
                using StreamProcessor.StreamProcessorBuilder streamProcessorBuilder = StreamProcessor.Builder();
                return streamProcessorBuilder
                    .Source(stream.Url)
                    .Region(stream.Region)
                    .Name(stream.Name)
                    .EventCallback(EventHandlerCallback)
                    .StatusChangeCallback
                    (
                        (stream1, status) =>
                        {
                            Console.WriteLine("Stream \"" + stream1.Name + "\" (" + stream1.SessionId
                                + ") status changed to \"" + status + "\"");

                            if (status == StreamProcessorStatus.Finished)
                            {
                                Console.WriteLine("STREAM PROCESSING HAS FINISHED in stream: \"" + stream1.Name
                                    + "\"! If the program is still running and you want to stop it, please press 'q' then 'Enter'");
                            }
                        }
                    )
                    .Anpr(anpr)
                    .Mmr(mmr)
                    .Roi(new List<Carmen.Point>
                    {
                        new Carmen.Point(0.03, 0.01),
                        new Carmen.Point(0.93, 0.04),
                        new Carmen.Point(0.98, 0.96),
                        new Carmen.Point(0.05, 0.97)
                    })
                    .AutoReconnect(true)
                    .Build();
                }
            ).ToList();

            Console.WriteLine("Please press 'q' then 'Enter' to stop stream processing.");

            // START STREAM PROCESSORS
            foreach (var stream in streams)
            {
                stream.Start();
            }

            // STREAM PROCESSORS ARE NOW RUNNING ASYNCHRONOUSLY

            // WAIT FOR KEY 'q' FROM STANDARD INPUT
            while (Console.ReadLine() != "q");

            // STOP STREAM PROCESSORS
            foreach (var stream in streams)
            {
                stream.Stop();
                stream.Dispose();
            }
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
        }

        Console.WriteLine("Please press a key to exit the program.");
        Console.ReadKey();
    }
}

File: sdk_samples/samples/C/04_multi_stream/04_multi_stream.c
/**
 * CARMEN Video SDK
 *
 * @category    C Sample
 * @package     CARMEN Video C Sample
 *
 * @copyright   2024 Adaptive Recognition
 * @licence     https://adaptiverecognition.com/eula/
 *
 * This sample shows how you can set up a process with CARMEN Video SDK that
 * recognizes license plates (ANPR) and vehicle attributes (MMR) on multiple video streams.
 * There are "TODO" comments in the file where the code must be modified.
 */

#include <stdio.h>
#include <string.h>
#include <inttypes.h>

#include "utils.h"

#include <carmen/Event.h>
#include <carmen/AnprBuilder.h>
#include <carmen/MmrBuilder.h>
#include <carmen/StreamProcessorBuilder.h>
#include <carmen/CmError.h>

typedef struct StreamConfig {
    const char* url;
    const char* region;
    const char* name;
} StreamConfig;

//TODO: change them for valid streams and engines
#define NUM_STREAMPROCESSORS 2
StreamConfig streamConfigs[NUM_STREAMPROCESSORS] = {
        {"http://192.168.6.50:9901/video.mjpeg",         "eur", "Stream 1"},
        {"file:///C:\\Program Files\\videos\\video.mp4", "nam", "Stream 2"}
};


//TODO: EventCallback is called from a StreamProcessor instance's dedicated "EventCallback executor" thread.
// If you are using the same resource in multiple StreamProcessors' EventCallback
// you may have to protect against race conditions
void onEventCallback(CM_EVENT* e, void* userdata) {

    printf("------------------------------------------------------\n");
    printf("Channel name: %s\n", e->channelName);
    printf("Unix timestamp: %" PRId64 " ms\n", e->timestamp);

    CM_VEHICLE* vehicle = e->vehicle;
    CM_PLATE* plate = vehicle->plate;

    printf("Plate text: %s\n", plate->text);
    printf("Country: %s\n", plate->country);

    printf("Make: %s\n", vehicle->mmrData->make);
    printf("Model: %s\n", vehicle->mmrData->model);
    printf("Color: ");
    printOptionalColorRGB(&vehicle->mmrData->color);
    printf("\n");
    printf("Category: %s\n", vehicle->mmrData->category);

    printf("\n");
    fflush(stdout);

    cm_event_free(e);
}

void onStatusChangeCallback(CM_STREAM_PROCESSOR handle, enum CM_StreamProcessorStatus status, void* userdata){
    commonStatusCallback(handle, status);
};

int main(int argc, char** argv) {
    // INITIALIZE COMMON ANPR OBJECT
    CM_ANPR_BUILDER anprBuilder;
    memset(&anprBuilder, 0, sizeof(anprBuilder));
    cm_anprbuilder_create(&anprBuilder);
    cm_anprbuilder_set_type(anprBuilder, CM_ANPR_TYPE_LOCAL); // use CM_ANPR_TYPE_LOCAL_GO for CARMEN_GO engines
    cm_anprbuilder_set_local_concurrency_limit(anprBuilder,
                                               2); //TODO: change it to a value that matches the number of core licences you have

    CM_ANPR anpr;
    memset(&anpr, 0, sizeof(anpr));
    CHECK_CM_ERROR(cm_anprbuilder_build(anprBuilder, &anpr));
    cm_anprbuilder_free(anprBuilder);

    // INITIALIZE COMMON MMR (Make and Model Recognition) OBJECT
    CM_MMR_BUILDER mmrBuilder;
    memset(&mmrBuilder, 0, sizeof(mmrBuilder));
    cm_mmrbuilder_create(&mmrBuilder);
    cm_mmrbuilder_set_type(mmrBuilder, CM_MMR_TYPE_LOCAL);

    CM_MMR mmr;
    memset(&mmr, 0, sizeof(mmr));
    CHECK_CM_ERROR(cm_mmrbuilder_build(mmrBuilder, &mmr));
    cm_mmrbuilder_free(mmrBuilder);

    // BUILD STREAM PROCESSOR OBJECTS THAT USE THE COMMON ANPR AND MMR RESOURCES
    CM_STREAM_PROCESSOR streamProcessors[NUM_STREAMPROCESSORS];
    memset(streamProcessors, 0, sizeof(streamProcessors));

    for(int i = 0; i < NUM_STREAMPROCESSORS; ++i) {
        CM_STREAM_PROCESSOR_BUILDER builder;
        memset(&builder, 0, sizeof(builder));
        cm_streamprocessorbuilder_create(&builder);

        StreamConfig* curConfig = &streamConfigs[i];

        cm_streamprocessorbuilder_set_source(builder, curConfig->url);
        cm_streamprocessorbuilder_set_region(builder, curConfig->region);
        cm_streamprocessorbuilder_set_name(builder, curConfig->name);
        cm_streamprocessorbuilder_set_event_callback(builder, onEventCallback, NULL, NULL);
        cm_streamprocessorbuilder_set_status_change_callback(builder, onStatusChangeCallback, NULL, NULL);
        cm_streamprocessorbuilder_set_anpr(builder, anpr);
        cm_streamprocessorbuilder_set_mmr(builder, mmr);
        cm_streamprocessorbuilder_set_anpr_color_recognition(builder, CM_TRUE);
        cm_streamprocessorbuilder_set_mmr_color_recognition(builder, CM_TRUE);
        cm_streamprocessorbuilder_set_auto_reconnect(builder, CM_TRUE);

        CM_POINT points[] = {
                {0.0,  0.0},
                {0.95, 0.0},
                {0.95, 0.95},
                {0.0,  0.95}
        };
        cm_streamprocessorbuilder_set_roi(builder, points, 4);

        CHECK_CM_ERROR(cm_streamprocessorbuilder_build(builder, &streamProcessors[i]));
        cm_streamprocessorbuilder_free(builder);
    }

	printf("Please press 'q' then 'Enter' to stop stream processing.\n");

    // START STREAM PROCESSORS
    for(int i = 0; i < NUM_STREAMPROCESSORS; ++i) {
        CHECK_CM_ERROR(cm_streamprocessor_start(streamProcessors[i]));
    }

    // STREAM PROCESSORS ARE NOW RUNNING ASYNCHRONOUSLY

    // WAIT FOR KEY 'q' FROM STANDARD INPUT
    while(getc(stdin) != 'q');

    // STOP STREAM PROCESSORS
    for(int i = 0; i < NUM_STREAMPROCESSORS; ++i) {
        CM_STREAM_PROCESSOR* curStream = &streamProcessors[i];
        CHECK_CM_ERROR(cm_streamprocessor_stop(*curStream));
        cm_streamprocessor_free(*curStream);
    }

    // FREE COMMON ANPR AND MMR OBJECT
    cm_anpr_free(anpr);
    cm_mmr_free(mmr);

    return 0;
}
